// pureimage/dist/named_colors.js
/**
 * Enumeration containing popular colors
 */
export var NAMED_COLORS;
(function (NAMED_COLORS) {
    NAMED_COLORS[NAMED_COLORS["transparent"] = 0] = "transparent";
    NAMED_COLORS[NAMED_COLORS["aliceblue"] = 4042850303] = "aliceblue";
    NAMED_COLORS[NAMED_COLORS["antiquewhite"] = 4209760255] = "antiquewhite";
    NAMED_COLORS[NAMED_COLORS["aqua"] = 16777215] = "aqua";
    NAMED_COLORS[NAMED_COLORS["aquamarine"] = 2147472639] = "aquamarine";
    NAMED_COLORS[NAMED_COLORS["azure"] = 4043309055] = "azure";
    NAMED_COLORS[NAMED_COLORS["beige"] = 4126530815] = "beige";
    NAMED_COLORS[NAMED_COLORS["bisque"] = 4293182719] = "bisque";
    NAMED_COLORS[NAMED_COLORS["black"] = 255] = "black";
    NAMED_COLORS[NAMED_COLORS["blanchedalmond"] = 4293643775] = "blanchedalmond";
    NAMED_COLORS[NAMED_COLORS["blue"] = 65535] = "blue";
    NAMED_COLORS[NAMED_COLORS["blueviolet"] = 2318131967] = "blueviolet";
    NAMED_COLORS[NAMED_COLORS["brown"] = 2771004159] = "brown";
    NAMED_COLORS[NAMED_COLORS["burlywood"] = 3736635391] = "burlywood";
    NAMED_COLORS[NAMED_COLORS["cadetblue"] = 1604231423] = "cadetblue";
    NAMED_COLORS[NAMED_COLORS["chartreuse"] = 2147418367] = "chartreuse";
    NAMED_COLORS[NAMED_COLORS["chocolate"] = 3530104575] = "chocolate";
    NAMED_COLORS[NAMED_COLORS["coral"] = 4286533887] = "coral";
    NAMED_COLORS[NAMED_COLORS["cornflowerblue"] = 1687547391] = "cornflowerblue";
    NAMED_COLORS[NAMED_COLORS["cornsilk"] = 4294499583] = "cornsilk";
    NAMED_COLORS[NAMED_COLORS["crimson"] = 3692313855] = "crimson";
    NAMED_COLORS[NAMED_COLORS["cyan"] = 16777215] = "cyan";
    NAMED_COLORS[NAMED_COLORS["darkblue"] = 35839] = "darkblue";
    NAMED_COLORS[NAMED_COLORS["darkcyan"] = 9145343] = "darkcyan";
    NAMED_COLORS[NAMED_COLORS["darkgoldenrod"] = 3095792639] = "darkgoldenrod";
    NAMED_COLORS[NAMED_COLORS["darkgray"] = 2846468607] = "darkgray";
    NAMED_COLORS[NAMED_COLORS["darkgreen"] = 6553855] = "darkgreen";
    NAMED_COLORS[NAMED_COLORS["darkgrey"] = 2846468607] = "darkgrey";
    NAMED_COLORS[NAMED_COLORS["darkkhaki"] = 3182914559] = "darkkhaki";
    NAMED_COLORS[NAMED_COLORS["darkmagenta"] = 2332068863] = "darkmagenta";
    NAMED_COLORS[NAMED_COLORS["darkolivegreen"] = 1433087999] = "darkolivegreen";
    NAMED_COLORS[NAMED_COLORS["darkorange"] = 4287365375] = "darkorange";
    NAMED_COLORS[NAMED_COLORS["darkorchid"] = 2570243327] = "darkorchid";
    NAMED_COLORS[NAMED_COLORS["darkred"] = 2332033279] = "darkred";
    NAMED_COLORS[NAMED_COLORS["darksalmon"] = 3918953215] = "darksalmon";
    NAMED_COLORS[NAMED_COLORS["darkseagreen"] = 2411499519] = "darkseagreen";
    NAMED_COLORS[NAMED_COLORS["darkslateblue"] = 1211993087] = "darkslateblue";
    NAMED_COLORS[NAMED_COLORS["darkslategray"] = 793726975] = "darkslategray";
    NAMED_COLORS[NAMED_COLORS["darkslategrey"] = 793726975] = "darkslategrey";
    NAMED_COLORS[NAMED_COLORS["darkturquoise"] = 13554175] = "darkturquoise";
    NAMED_COLORS[NAMED_COLORS["darkviolet"] = 2483082239] = "darkviolet";
    NAMED_COLORS[NAMED_COLORS["deeppink"] = 4279538687] = "deeppink";
    NAMED_COLORS[NAMED_COLORS["deepskyblue"] = 12582911] = "deepskyblue";
    NAMED_COLORS[NAMED_COLORS["dimgray"] = 1768516095] = "dimgray";
    NAMED_COLORS[NAMED_COLORS["dimgrey"] = 1768516095] = "dimgrey";
    NAMED_COLORS[NAMED_COLORS["dodgerblue"] = 512819199] = "dodgerblue";
    NAMED_COLORS[NAMED_COLORS["firebrick"] = 2988581631] = "firebrick";
    NAMED_COLORS[NAMED_COLORS["floralwhite"] = 4294635775] = "floralwhite";
    NAMED_COLORS[NAMED_COLORS["forestgreen"] = 579543807] = "forestgreen";
    NAMED_COLORS[NAMED_COLORS["fuchsia"] = 4278255615] = "fuchsia";
    NAMED_COLORS[NAMED_COLORS["gainsboro"] = 3705462015] = "gainsboro";
    NAMED_COLORS[NAMED_COLORS["ghostwhite"] = 4177068031] = "ghostwhite";
    NAMED_COLORS[NAMED_COLORS["gold"] = 4292280575] = "gold";
    NAMED_COLORS[NAMED_COLORS["goldenrod"] = 3668254975] = "goldenrod";
    NAMED_COLORS[NAMED_COLORS["gray"] = 2155905279] = "gray";
    NAMED_COLORS[NAMED_COLORS["green"] = 8388863] = "green";
    NAMED_COLORS[NAMED_COLORS["greenyellow"] = 2919182335] = "greenyellow";
    NAMED_COLORS[NAMED_COLORS["grey"] = 2155905279] = "grey";
    NAMED_COLORS[NAMED_COLORS["honeydew"] = 4043305215] = "honeydew";
    NAMED_COLORS[NAMED_COLORS["hotpink"] = 4285117695] = "hotpink";
    NAMED_COLORS[NAMED_COLORS["indianred"] = 3445382399] = "indianred";
    NAMED_COLORS[NAMED_COLORS["indigo"] = 1258324735] = "indigo";
    NAMED_COLORS[NAMED_COLORS["ivory"] = 4294963455] = "ivory";
    NAMED_COLORS[NAMED_COLORS["khaki"] = 4041641215] = "khaki";
    NAMED_COLORS[NAMED_COLORS["lavender"] = 3873897215] = "lavender";
    NAMED_COLORS[NAMED_COLORS["lavenderblush"] = 4293981695] = "lavenderblush";
    NAMED_COLORS[NAMED_COLORS["lawngreen"] = 2096890111] = "lawngreen";
    NAMED_COLORS[NAMED_COLORS["lemonchiffon"] = 4294626815] = "lemonchiffon";
    NAMED_COLORS[NAMED_COLORS["lightblue"] = 2916673279] = "lightblue";
    NAMED_COLORS[NAMED_COLORS["lightcoral"] = 4034953471] = "lightcoral";
    NAMED_COLORS[NAMED_COLORS["lightcyan"] = 3774873599] = "lightcyan";
    NAMED_COLORS[NAMED_COLORS["lightgoldenrodyellow"] = 4210742015] = "lightgoldenrodyellow";
    NAMED_COLORS[NAMED_COLORS["lightgray"] = 3553874943] = "lightgray";
    NAMED_COLORS[NAMED_COLORS["lightgreen"] = 2431553791] = "lightgreen";
    NAMED_COLORS[NAMED_COLORS["lightgrey"] = 3553874943] = "lightgrey";
    NAMED_COLORS[NAMED_COLORS["lightpink"] = 4290167295] = "lightpink";
    NAMED_COLORS[NAMED_COLORS["lightsalmon"] = 4288707327] = "lightsalmon";
    NAMED_COLORS[NAMED_COLORS["lightseagreen"] = 548580095] = "lightseagreen";
    NAMED_COLORS[NAMED_COLORS["lightskyblue"] = 2278488831] = "lightskyblue";
    NAMED_COLORS[NAMED_COLORS["lightslategray"] = 2005441023] = "lightslategray";
    NAMED_COLORS[NAMED_COLORS["lightslategrey"] = 2005441023] = "lightslategrey";
    NAMED_COLORS[NAMED_COLORS["lightsteelblue"] = 2965692159] = "lightsteelblue";
    NAMED_COLORS[NAMED_COLORS["lightyellow"] = 4294959359] = "lightyellow";
    NAMED_COLORS[NAMED_COLORS["lime"] = 16711935] = "lime";
    NAMED_COLORS[NAMED_COLORS["limegreen"] = 852308735] = "limegreen";
    NAMED_COLORS[NAMED_COLORS["linen"] = 4210091775] = "linen";
    NAMED_COLORS[NAMED_COLORS["magenta"] = 4278255615] = "magenta";
    NAMED_COLORS[NAMED_COLORS["maroon"] = 2147483903] = "maroon";
    NAMED_COLORS[NAMED_COLORS["mediumaquamarine"] = 1724754687] = "mediumaquamarine";
    NAMED_COLORS[NAMED_COLORS["mediumblue"] = 52735] = "mediumblue";
    NAMED_COLORS[NAMED_COLORS["mediumorchid"] = 3126187007] = "mediumorchid";
    NAMED_COLORS[NAMED_COLORS["mediumpurple"] = 2473647103] = "mediumpurple";
    NAMED_COLORS[NAMED_COLORS["mediumseagreen"] = 1018393087] = "mediumseagreen";
    NAMED_COLORS[NAMED_COLORS["mediumslateblue"] = 2070474495] = "mediumslateblue";
    NAMED_COLORS[NAMED_COLORS["mediumspringgreen"] = 16423679] = "mediumspringgreen";
    NAMED_COLORS[NAMED_COLORS["mediumturquoise"] = 1221709055] = "mediumturquoise";
    NAMED_COLORS[NAMED_COLORS["mediumvioletred"] = 3340076543] = "mediumvioletred";
    NAMED_COLORS[NAMED_COLORS["midnightblue"] = 421097727] = "midnightblue";
    NAMED_COLORS[NAMED_COLORS["mintcream"] = 4127193855] = "mintcream";
    NAMED_COLORS[NAMED_COLORS["mistyrose"] = 4293190143] = "mistyrose";
    NAMED_COLORS[NAMED_COLORS["moccasin"] = 4293178879] = "moccasin";
    NAMED_COLORS[NAMED_COLORS["navajowhite"] = 4292783615] = "navajowhite";
    NAMED_COLORS[NAMED_COLORS["navy"] = 33023] = "navy";
    NAMED_COLORS[NAMED_COLORS["oldlace"] = 4260751103] = "oldlace";
    NAMED_COLORS[NAMED_COLORS["olive"] = 2155872511] = "olive";
    NAMED_COLORS[NAMED_COLORS["olivedrab"] = 1804477439] = "olivedrab";
    NAMED_COLORS[NAMED_COLORS["orange"] = 4289003775] = "orange";
    NAMED_COLORS[NAMED_COLORS["orangered"] = 4282712319] = "orangered";
    NAMED_COLORS[NAMED_COLORS["orchid"] = 3664828159] = "orchid";
    NAMED_COLORS[NAMED_COLORS["palegoldenrod"] = 4008225535] = "palegoldenrod";
    NAMED_COLORS[NAMED_COLORS["palegreen"] = 2566625535] = "palegreen";
    NAMED_COLORS[NAMED_COLORS["paleturquoise"] = 2951671551] = "paleturquoise";
    NAMED_COLORS[NAMED_COLORS["palevioletred"] = 3681588223] = "palevioletred";
    NAMED_COLORS[NAMED_COLORS["papayawhip"] = 4293907967] = "papayawhip";
    NAMED_COLORS[NAMED_COLORS["peachpuff"] = 4292524543] = "peachpuff";
    NAMED_COLORS[NAMED_COLORS["peru"] = 3448061951] = "peru";
    NAMED_COLORS[NAMED_COLORS["pink"] = 4290825215] = "pink";
    NAMED_COLORS[NAMED_COLORS["plum"] = 3718307327] = "plum";
    NAMED_COLORS[NAMED_COLORS["powderblue"] = 2967529215] = "powderblue";
    NAMED_COLORS[NAMED_COLORS["purple"] = 2147516671] = "purple";
    NAMED_COLORS[NAMED_COLORS["rebeccapurple"] = 1714657791] = "rebeccapurple";
    NAMED_COLORS[NAMED_COLORS["red"] = 4278190335] = "red";
    NAMED_COLORS[NAMED_COLORS["rosybrown"] = 3163525119] = "rosybrown";
    NAMED_COLORS[NAMED_COLORS["royalblue"] = 1097458175] = "royalblue";
    NAMED_COLORS[NAMED_COLORS["saddlebrown"] = 2336560127] = "saddlebrown";
    NAMED_COLORS[NAMED_COLORS["salmon"] = 4202722047] = "salmon";
    NAMED_COLORS[NAMED_COLORS["sandybrown"] = 4104413439] = "sandybrown";
    NAMED_COLORS[NAMED_COLORS["seagreen"] = 780883967] = "seagreen";
    NAMED_COLORS[NAMED_COLORS["seashell"] = 4294307583] = "seashell";
    NAMED_COLORS[NAMED_COLORS["sienna"] = 2689740287] = "sienna";
    NAMED_COLORS[NAMED_COLORS["silver"] = 3233857791] = "silver";
    NAMED_COLORS[NAMED_COLORS["skyblue"] = 2278484991] = "skyblue";
    NAMED_COLORS[NAMED_COLORS["slateblue"] = 1784335871] = "slateblue";
    NAMED_COLORS[NAMED_COLORS["slategray"] = 1887473919] = "slategray";
    NAMED_COLORS[NAMED_COLORS["slategrey"] = 1887473919] = "slategrey";
    NAMED_COLORS[NAMED_COLORS["snow"] = 4294638335] = "snow";
    NAMED_COLORS[NAMED_COLORS["springgreen"] = 16744447] = "springgreen";
    NAMED_COLORS[NAMED_COLORS["steelblue"] = 1182971135] = "steelblue";
    NAMED_COLORS[NAMED_COLORS["tan"] = 3535047935] = "tan";
    NAMED_COLORS[NAMED_COLORS["teal"] = 8421631] = "teal";
    NAMED_COLORS[NAMED_COLORS["thistle"] = 3636451583] = "thistle";
    NAMED_COLORS[NAMED_COLORS["tomato"] = 4284696575] = "tomato";
    NAMED_COLORS[NAMED_COLORS["turquoise"] = 1088475391] = "turquoise";
    NAMED_COLORS[NAMED_COLORS["violet"] = 4001558271] = "violet";
    NAMED_COLORS[NAMED_COLORS["wheat"] = 4125012991] = "wheat";
    NAMED_COLORS[NAMED_COLORS["white"] = 4294967295] = "white";
    NAMED_COLORS[NAMED_COLORS["whitesmoke"] = 4126537215] = "whitesmoke";
    NAMED_COLORS[NAMED_COLORS["yellow"] = 4294902015] = "yellow";
    NAMED_COLORS[NAMED_COLORS["yellowgreen"] = 2597139199] = "yellowgreen";
})(NAMED_COLORS || (NAMED_COLORS = {}));
export const TRANSPARENT_BLACK = 0x00000000;
export const OPAQUE_BLACK = 0x000000ff;
